### PR TITLE
State Canonicalization and Hashing

### PR DESCRIPTION
**Pull request by jacobhaven**
r? @maxtaco

### CODE REVIEW

**Comment by jacobhaven** [comment]
It was simplest to always sort the items in  `ObjFactory#MakeState` rather than assuming that all `StateMachine`s correctly sort their items (`SQLEngine` does, `MemEngine` doesn't).

**Comment by maxtaco** [comment]
Are there any sort of tests that make sense here?

**Comment by jacobhaven** [comment]
@maxtaco creating a slice of items items, hashing them, shuffling the slice and ensuring the hash is the same. I haven't gotten around to writing the boilerplate to generate that slice though...

File: protocol/gregor1/factory.go [hunk 1]
```diff
@@ -148,15 +149,25 @@ func (o ObjFactory) MakeStateSyncMessage(uid gregor.UID, msgid gregor.MsgID, dev
 	}, nil
 }
 
+type itemSlice []ItemAndMetadata
+
+func (its itemSlice) Len() int      { return len(its) }
+func (its itemSlice) Swap(i, j int) { its[i], its[j] = its[j], its[i] }
+func (its itemSlice) Less(i, j int) bool {
+	mI, mJ := its[i].Metadata(), its[j].Metadata()
+	return mI == nil || (mJ != nil && mI.CTime().Before(mJ.CTime()))
```
Comment (jacobhaven) [inline protocol/gregor1/factory.go]: Needs a tie-breaker if the ctimes are equal. `MsgID`?

Comment (maxtaco) [inline protocol/gregor1/factory.go]: yes, good idea.

**Comment by maxtaco** [comment]
:+1:

**Pull request merged**